### PR TITLE
feat: Make SecretStr hashable

### LINKED ISSUE #4387: Hashable SecretStr
### Initial Checks

- [X] I have searched Google & GitHub for similar requests and couldn't find anything
- [X] I have read and followed [the docs](https://pydantic-docs.helpmanual.io) and still think this feature is missing

### Description

I'd like to see instances of the pydantic class `SecretStr` to be hashable like instances of the `str` class are hashable.

I use it mainly in instances of `BaseSettings` and during startup of a server application, I add settings classes which I can successfully construct in a set. This gives me an indication of missing configuration values.
But models containing fields annotated with `SecretStr` currently cannot be added to a set as Python complains about the `SecretStr` type: `TypeError: unhashable type: 'SecretStr'`

Example snippet:

```py
import pydantic

class M(pydantic.BaseModel):
    s: pydantic.SecretStr = pydantic.SecretStr("secret")

    def __hash__(self):
        return hash((type(self),) + tuple(self.__dict__.values()))

def test_can_hash():
    assert hash(M())
```

### Affected Components

- [ ] [Compatibility between releases](https://pydantic-docs.helpmanual.io/changelog/)
- [ ] [Data validation/parsing](https://pydantic-docs.helpmanual.io/usage/models/#basic-model-usage)
- [ ] [Data serialization](https://pydantic-docs.helpmanual.io/usage/exporting_models/) - `.dict()` and `.json()`
- [ ] [JSON Schema](https://pydantic-docs.helpmanual.io/usage/schema/)
- [ ] [Dataclasses](https://pydantic-docs.helpmanual.io/usage/dataclasses/)
- [ ] [Model Config](https://pydantic-docs.helpmanual.io/usage/model_config/)
- [ ] [Field Types](https://pydantic-docs.helpmanual.io/usage/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://pydantic-docs.helpmanual.io/usage/validation_decorator/)
- [ ] [Generic Models](https://pydantic-docs.helpmanual.io/usage/models/#generic-models)
- [X] [Other Model behaviour](https://pydantic-docs.helpmanual.io/usage/models/) - `construct()`, pickling, private attributes, ORM mode
- [ ] [Settings Management](https://pydantic-docs.helpmanual.io/usage/settings/)
- [ ] [Plugins](https://pydantic-docs.helpmanual.io/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

### PR DESCRIPTION
**Pull request by chbndrhnns**
## Change Summary

Makes `SecretStr` hashable, cf. https://github.com/pydantic/pydantic/issues/4387

## Related issue number

fix #4387 

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/pydantic/pydantic/blob/master/changes/README.md) for details)
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

**Comment by chbndrhnns** [comment]
please review

File: tests/test_types.py [hunk 1]
```diff
@@ -2650,6 +2650,10 @@ class Foobar(BaseModel):
     assert m.password.get_secret_value() == '1234'
 
 
+def test_secretstr_is_hashable():
+    assert hash(SecretStr('secret'))
```
Comment (hramezani) [inline tests/test_types.py:2654]: maybe:

```suggestion
    assert type(hash(SecretStr('secret'))) is int
```

**Comment by hramezani** [comment]
please update

**Comment by chbndrhnns** [comment]
Do I need another please review here? :)

File: pydantic/types.py [hunk 2]
```diff
@@ -875,6 +875,9 @@ def __repr__(self) -> str:
     def __len__(self) -> int:
         return len(self._secret_value)
 
+    def __hash__(self) -> int:
```
Comment (samuelcolvin) [inline pydantic/types.py:878]: I think you can move this onto `SecretField` and then both `SecretStr` and `SecretBytes` should be hashable.

You can then also add a test for `assert type(hash(SecretBytes('secret'))) is int`

**Review by samuelcolvin — COMMENTED** [review]
otherwise LGTM.

Please update.

File: pydantic/types.py (hunk 2 above)
Comment (hramezani) [reply pydantic/types.py:878]: Also, you need to update your change entry as well.

File: changes/4387-chbndrhnns.md [hunk 3]
```diff
@@ -0,0 +1 @@
+Make `SecretStr` hashable
```
Comment (hramezani) [inline changes/4387-chbndrhnns.md:1]:
```suggestion
Make `SecretStr` and `SecretBytes` hashable
```

**Comment by chbndrhnns** [comment]
Please review

**Comment by samuelcolvin** [comment]
Great, thanks so much.

**Pull request merged**